cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+
+static void uzfs_zvol_io_ack_sender(void *arg);
+
+static int
+create_and_bind(const char *port)
+{
+	int s, sfd;
+	struct addrinfo hints = {0, };
+	struct addrinfo *rp, *result = NULL;
+
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_flags = AI_PASSIVE;
+
+	s = getaddrinfo(NULL, port, &hints, &result);
+	if (s != 0) {
+		ZREPL_ERRLOG("getaddrinfo failed with error: %d\n", errno);
+		return (-1);
+	}
+
+	for (rp = result; rp != NULL; rp = rp->ai_next) {
+		sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+		if (sfd == -1) {
+			continue;
+		}
+
+		s = bind(sfd, rp->ai_addr, rp->ai_addrlen);
+		if (s == 0) {
+			/* We managed to bind successfully! */
+			ZREPL_LOG("bind is successful\n");
+			break;
+		}
+		close(sfd);
+	}
+
+	freeaddrinfo(result);
+	if (rp == NULL) {
+		ZREPL_ERRLOG("bind failed with err:%d\n", errno);
+		return (-1);
+	}
+
+	return (sfd);
+}
+
+static int
+make_socket_non_blocking(int sfd)
+{
+	int flags, s;
+
+	flags = fcntl(sfd, F_GETFL, 0);
+	if (flags == -1) {
+		ZREPL_ERRLOG("fcntl() failed errno:%d\n", errno);
+		return (-1);
+	}
+
+	flags |= O_NONBLOCK;
+	s = fcntl(sfd, F_SETFL, flags);
+	if (s == -1) {
+		ZREPL_ERRLOG("fcntl() failed errno:%d\n", errno);
+		return (-1);
+	}
+	return (0);
+}
+
+static int
+uzfs_zvol_get_ip(char *host)
+{
+	struct ifaddrs *ifaddr, *ifa;
+	int family, s, n;
+
+	if (getifaddrs(&ifaddr) == -1) {
+		ZREPL_ERRLOG("getifaddrs() failed errno:%d\n", errno);
+		return (-1);
+	}
+
+	/*
+	 * Walk through linked list, maintaining head
+	 * pointer so we can free list later
+	 */
+
+	for (ifa = ifaddr, n = 0; ifa != NULL; ifa = ifa->ifa_next, n++) {
+		if (ifa->ifa_addr == NULL) {
+			continue;
+		}
+
+		family = ifa->ifa_addr->sa_family;
+
+		if (family == AF_INET || family == AF_INET6) {
+			s = getnameinfo(ifa->ifa_addr, (family == AF_INET) ?
+			    sizeof (struct sockaddr_in) :
+			    sizeof (struct sockaddr_in6),
+			    host, NI_MAXHOST,
+			    NULL, 0, NI_NUMERICHOST);
+			if (s != 0) {
+				ZREPL_ERRLOG("getnameinfo() failed: %d\n",
+				    errno);
+				s = -1;
+				goto exit;
+			}
+
+			if (family == AF_INET) {
+				if (strcmp(host, "127.0.0.1") == 0) {
+					continue;
+				}
+				printf("IP address: %s\n", host);
+				break;
+			}
+		}
+	}
+exit:
+	freeifaddrs(ifaddr);
+	return (s);
+}
+/*
+ * Allocate zio command along with
+ * buffer needed for IO completion.
+ */
+static zvol_io_cmd_t *
+zio_cmd_alloc(zvol_io_hdr_t *hdr, int fd)
+{
+	zvol_io_cmd_t *zio_cmd = kmem_zalloc(
+	    sizeof (zvol_io_cmd_t), KM_SLEEP);
+
+	bcopy(hdr, &zio_cmd->hdr, sizeof (zio_cmd->hdr));
+	if ((hdr->opcode == ZVOL_OPCODE_READ) ||
+	    (hdr->opcode == ZVOL_OPCODE_WRITE) ||
+	    (hdr->opcode == ZVOL_OPCODE_HANDSHAKE)) {
+		zio_cmd->buf = kmem_zalloc(sizeof (char) * hdr->len, KM_SLEEP);
+	}
+
+	zio_cmd->conn = fd;
+	return (zio_cmd);
+}
+
+/*
+ * Free zio command along with buffer.
+ */
+static void
+zio_cmd_free(zvol_io_cmd_t **cmd, zvol_op_code_t opcode)
```

Comment:
Opcode is part of cmd->hdr, hdr is allocated from stack then we use hdr info to allocate cmd memory and buffer for read/write of size equal to hdr->len. Once we allocated cmd, we bcopy hdr to cmd->hdr.

Reply:
ok, but if I understand correctly since hdr is part of cmd (no matter how it was constructed) it contains the opcode and we don't need to pass it explicitly as a parameter to this function.